File: 프로그래머스/1/42889. 실패율/실패율.py
from collections import Counter
def solution(N, stages):
    stages_cnt = Counter(stages)

    result = []
    acc = 0
    total = len(stages)
    for i in range(1, N+1):
        if total - acc == 0:
            fail_rate = 0
        else:
            fail_rate = stages_cnt[i] / (total - acc)
        acc += stages_cnt[i]
        result.append([fail_rate, i])
        
    answer = [row[1] for row in sorted(result, key = lambda x: x[0], reverse=True)]
    return answer

        

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    const stagesCnt = new Map();
    for(const stage of stages){
        stagesCnt.set(stage, (stagesCnt.get(stage) || 0) + 1);
    }
    
    const result = [];
    const total = stages.length;
    let acc = 0;
    
    for(let i = 1 ; i <= N ; i++){
        const current = stagesCnt.get(i) || 0;
        const remaining = total - acc;
        const failRate = remaining === 0 ? 0 : current / remaining;
        acc += current;
        result.push([failRate, i]);
    }
    
    result.sort((a, b) => {
        if (b[0] === a[0]) {
            return a[1] - b[1]; // 실패율 같으면 스테이지 번호 낮은 순
        }
        return b[0] - a[0];
    });
    
    return result.map(item => item[1]);
}